File: experiments/proto/layer_shared_bitset_audit.py
#!/usr/bin/env python3
"""SHA-pinned shared-F4 resume audit using an independent native bitset reader.

Only data/logs artifacts are written. The original catalogue is never opened.
Production mode checks a strictly extended prefix, both physical backups, local
record semantics, and all current-prefix alias targets. Same-prefix differential
qualification is explicit TEST_ONLY_PASS and cannot seed a production chain.
"""
from __future__ import annotations
import argparse
import csv
import hashlib
import io
import json
from pathlib import Path
import re
import struct
import subprocess
import time

from layer_shared_checkpoint_audit import MASK, PINNED_SHA, checksum, fields, header, repair_key
from layer_shared_resume_audit import HardBounds, bounded_bytes, require

AUDIT_TYPE = 'shared-f4-native-bitset-resume-v1'
LEGACY_SCOPES = {
    'committed chunk integrity and physical-backup audit; not independent F4 recomputation',
    'independent immutable-file/resume/backup audit, not independent F4 recalculation',
    'streaming immutable resume/backup integrity audit; not independent F4 recalculation',
}
N, CHUNK = 903398621, 25000
MAX_FILES = (N+CHUNK-1)//CHUNK+1
MAX_BYTES = 12*N+256*MAX_FILES
PATTERN = re.compile(r'chunk-[0-9a-f]{16}\.bin\Z')
SHA = re.compile(r'[A-F0-9]{64}\Z')
LOG_ROOT = Path(__file__).resolve().parents[2]/'data'/'logs'


def digest(raw: bytes) -> str:
    return hashlib.sha256(raw).hexdigest().upper()


def inventory(directory: Path) -> list[str]:
    require(directory.is_dir(), 'missing audit input directory: '+str(directory))
    names = []
    for path in directory.iterdir():
        if path.name == 'manifest.bin' or PATTERN.fullmatch(path.name):
            names.append(path.name)
            require(len(names) <= MAX_FILES, 'committed inventory exceeds complete domain')
    return sorted(names)  # .tmp is never opened and cannot count as committed.


def receipt(directory: Path, names: list[str]) -> tuple[dict, str]:
    raw = bounded_bytes(directory/'receipt.csv', 16 << 20)
    rows = {}
    for row in csv.DictReader(io.StringIO(raw.decode('utf-8-sig'))):
        name = row.get('Name', '')
        require(name not in rows and (name == 'manifest.bin' or PATTERN.fullmatch(name)),
                'duplicate or malformed receipt name')
        require(SHA.fullmatch(row.get('SHA256', '')) is not None, 'invalid receipt SHA')
        require(re.fullmatch(r'[0-9]{1,12}', row.get('Bytes', '')) is not None, 'invalid receipt byte count')
        rows[name] = dict(name=name, bytes=int(row['Bytes']), sha256=row['SHA256'])
    require(sorted(rows) == names, 'receipt inventory mismatch: '+str(directory))
    return rows, digest(raw)


def predecessor(path: Path, expected_sha: str) -> tuple[dict, dict]:
    raw = bounded_bytes(path, 32 << 20)
    require(digest(raw) == expected_sha, 'previous audit report SHA mismatch')
    previous = json.loads(raw)
    require(previous.get('status') == 'PASS', 'previous report is not a production PASS')
    require(previous.get('prefix_snapshot', False) is False and
            previous.get('backup_snapshot_test', False) is False and
            previous.get('qualification_replay', False) is False and
            previous.get('protocol_test', False) is False,
            'a production audit chain cannot inherit a snapshot/test report')
    if 'audit_type' in previous:
        require(previous['audit_type'] == AUDIT_TYPE and
                previous.get('prefix_snapshot') is False and
                previous.get('backup_snapshot_test') is False and
                previous.get('qualification_replay') is False and
                previous.get('protocol_test') is False and
                previous.get('global_alias_closure_checked') is True,
                'unrecognized or incomplete explicit production audit type')
    else:
        require(previous.get('scope') in LEGACY_SCOPES, 'unrecognized legacy audit scope')
    require(previous.get('source_sha256_lineage', PINNED_SHA) == PINNED_SHA,
            'previous source lineage differs')
    rows = previous['files_audited']
    require(isinstance(rows, list) and 1 < len(rows) <= MAX_FILES, 'invalid previous file list')
    mapped = {}
    for row in rows:
        name = row['name']
        require(name not in mapped and (name == 'manifest.bin' or PATTERN.fullmatch(name)) and
                isinstance(row['bytes'], int) and not isinstance(row['bytes'], bool) and
                0 < row['bytes'] <= 300256 and SHA.fullmatch(row['sha256']),
                'invalid or duplicate previous file entry')
        mapped[name] = row
    return previous, mapped


def write_json(path: Path, value: object) -> None:
    with path.open('x', encoding='utf-8', newline='\n') as stream:
        json.dump(value, stream, indent=2)
        stream.write('\n')


def audit(args: argparse.Namespace, guard: HardBounds) -> dict:
    previous, old = predecessor(args.previous_report, args.previous_sha256)
    prior = int(previous['closed_prefix'])
    prefix = args.expected_prefix
    require(0 < prior <= prefix <= N and (prior % CHUNK == 0 or prior == N), 'invalid predecessor prefix')
    require(prior == prefix if args.qualification_replay else prior < prefix,
            'production requires a strict extension; qualification requires the same prefix')
    names = [f'chunk-{i:016x}.bin' for i in range(0,prefix,CHUNK)]+['manifest.bin']
    old_names = [f'chunk-{i:016x}.bin' for i in range(0,prior,CHUNK)]+['manifest.bin']
    require(sorted(old) == old_names, 'previous report is not an exact contiguous prefix')
    require(inventory(args.namespace) == names and inventory(args.backup_after) == names,
            'current/after inventory is not the exact required prefix')
    require(args.namespace.resolve().drive.casefold() != args.backup_after.resolve().drive.casefold(),
            'current and physical after backup must be on separate Windows volumes')
    after, after_sha = receipt(args.backup_after, names)
    before_sha = None
    if not args.qualification_replay:
        require(args.backup_before is not None, 'production requires a before backup')
        require(args.backup_before.resolve() != args.backup_after.resolve() and
                args.namespace.resolve().drive.casefold() != args.backup_before.resolve().drive.casefold(),
                'before backup must be distinct and on a separate volume')
        require(inventory(args.backup_before) == old_names, 'before inventory differs from predecessor')
        before, before_sha = receipt(args.backup_before, old_names)
        require(all(before[name] == {k:old[name][k] for k in ('name','bytes','sha256')} for name in old_names),
                'before receipt differs from SHA-pinned predecessor')
    for name in old_names:
        require(after[name] == {k:old[name][k] for k in ('name','bytes','sha256')},
                'old after receipt differs from SHA-pinned predecessor')
    total_bytes = sum(row['bytes'] for row in after.values())
    require(total_bytes <= args.max_bytes, 'declared committed-byte budget exceeded')
    raw_manifest = bounded_bytes(args.namespace/'manifest.bin', 256)
    manifest = header(raw_manifest)
    require(manifest['magic'] == 'SFR4MT01' and manifest['entries'] == N and manifest['chunk_size'] == CHUNK and
            manifest['source_header_hash'] == 10599038447932118899,
            'unexpected production manifest/domain/source header')
    key, stab = repair_key()
    repair_hash = checksum(struct.pack('<I',stab),checksum(struct.pack('<12H',*key),
                           checksum(b'L4-stab12-order3-witness-v1')))
    require(manifest['repair_hash'] == repair_hash, 'independent repair hash differs')
    mode = 'replay' if args.qualification_replay else 'resume'
    plan_text = f'SHARED_F4_BITSET_PLAN_V1\n{prefix} {prior} {len(names)} {args.max_bytes} {mode}\n'
    for name in names:
        row = after[name]
        plan_text += f"{name}\t{row['bytes']}\t{row['sha256']}\t{int(not args.qualification_replay and name in old)}\n"
    plan_path = args.output.with_suffix('.plan.txt')
    with plan_path.open('x',encoding='ascii',newline='\n') as stream:
        stream.write(plan_text)
    plan_sha = digest(plan_text.encode('ascii'))
    worker_sha = digest(bounded_bytes(args.worker, 16 << 20))
    seconds = int(args.maxseconds-(time.monotonic()-guard.started))-1
    require(seconds > 0, 'no bounded worker time remains')
    command = [str(args.worker.resolve()), 'plan='+str(plan_path.resolve()), 'plan_sha256='+plan_sha,
               'current='+str(args.namespace.resolve()), 'after='+str(args.backup_after.resolve()),
               'before='+str((args.backup_before or args.backup_after).resolve()),
               'maxseconds='+str(seconds), 'maxrssmib=1024']
    worker_stdout = args.output.with_suffix('.worker.jsonl')
    worker_stderr = args.output.with_suffix('.worker.stderr.txt')
    with worker_stdout.open('xb') as out, worker_stderr.open('xb') as err:
        completed = subprocess.run(command,stdout=out,stderr=err,timeout=seconds+0.5,check=False)
    require(completed.returncode == 0, 'native audit worker failed; see '+str(worker_stderr))
    require(digest(bounded_bytes(args.worker,16<<20)) == worker_sha, 'worker executable changed during audit')
    require(digest(bounded_bytes(plan_path,16<<20)) == plan_sha, 'worker plan changed during audit')
    require(bounded_bytes(worker_stderr,1<<20) == b'', 'native worker stderr is not empty')
    parsed = [json.loads(line) for line in bounded_bytes(worker_stdout,32<<20).decode('ascii').splitlines()]
    require(len(parsed) == len(names)+1 and parsed[-1].get('worker_status') == 'PASS', 'incomplete worker certificate')
    reports, totals = parsed[:-1], parsed[-1]
    require([r['name'] for r in reports] == names, 'worker file inventory differs')
    for row in reports:
        require(all(row[k] == after[row['name']][k] for k in ('name','bytes','sha256')), 'worker file digest differs')
        if row['name'] in old:
            require(all(row[k] == old[row['name']][k] for k in row if k in old[row['name']]),
                    'previous independently decoded file record differs')
    require(totals['closed_prefix'] == prefix and totals['total_bytes'] == total_bytes and
            totals['prior_live_records'] == previous['live_records'] and
            totals['prior_closed_representatives'] == previous['closed_representatives'] and
            totals['prior_value_checksum'] == previous.get('value_checksum',previous.get('F4_checksum_mod2_64')),
            'worker/previous counters disagree')
    new_rows = [r for r in reports if r.get('begin',-1) >= prior]
    summary = previous.get('engine_summary',{})
    stdout_sha = None
    if args.qualification_replay:
        for key_name in ('live_records','closed_representatives','aliases_to_closed_representatives',
                         'aliases_to_future_uncomputed_representatives'):
            require(totals[key_name] == previous[key_name], 'qualification differential counter differs: '+key_name)
        require(totals['holes'] == previous.get('holes',prefix-previous['live_records']), 'qualification holes differ')
    else:
        require(args.stdout is not None, 'production audit needs the terminal stdout log')
        raw = bounded_bytes(args.stdout,32<<20)
        stdout_sha = digest(raw)
        lines = raw.decode('utf-8').splitlines()
        summaries = [fields(line) for line in lines if line.startswith('SUMMARY ')]
        resumed = [fields(line) for line in lines if line.startswith('RESUMED ')]
        fresh = [fields(line) for line in lines if line.startswith('CLOSED_CHUNK ')]
        require(len(summaries) == len(resumed) == 1 and len(fresh) == len(new_rows), 'terminal log inventory differs')
        summary, resume = summaries[0], resumed[0]
        require((int(resume['chunks']),int(resume['closed_prefix']),int(resume['closed_representatives'])) ==
                (len(old_names)-1,prior,previous['closed_representatives']), 'logged resumed state differs')
        for row, logged in zip(new_rows,fresh):
            require(all(row[k] == int(logged[k]) for k in ('begin','count','live','representatives')) and
                    logged['closed_prefix'] == f"{row['begin']+row['count']}/{N}", 'new chunk log differs')
        expected = dict(new_chunks=len(new_rows),new_indices=prefix-prior,closed_prefix=prefix,total_entries=N,
                        closed_representatives=totals['closed_representatives'],F4_checksum_mod2_64=totals['value_checksum'])
        require(all(int(summary[k]) == value for k,value in expected.items()), 'terminal summary differs')
        require(summary.get('N6') == 'NOT_COMPUTED' and
                summary.get('status') == ('CLOSED_F4_CATALOGUE' if prefix == N else 'INCOMPLETE_RESUMABLE'),
                'terminal F4 scope marker differs')
        require(bounded_bytes(args.stdout.parent/'stderr.log',1<<20) == b'', 'producer stderr is not empty')
    require(inventory(args.namespace) == names and inventory(args.backup_after) == names,
            'committed current/after inventory changed during audit')
    require(receipt(args.backup_after,names)[1] == after_sha, 'after receipt changed during audit')
    if not args.qualification_replay:
        require(inventory(args.backup_before) == old_names and receipt(args.backup_before,old_names)[1] == before_sha,
                'before backup inventory/receipt changed during audit')
    return dict(status='TEST_ONLY_PASS' if args.qualification_replay or args.protocol_test else 'PASS',audit_type=AUDIT_TYPE,
        scope='independent native two-pass byte/backup/alias audit; not independent F4 recomputation',
        storage_integrity_scope='content/provenance at the verified reads, not a simultaneous or permanently immutable namespace snapshot',
        qualification_replay=args.qualification_replay,protocol_test=args.protocol_test,
        prefix_snapshot=False,backup_snapshot_test=False,
        namespace=str(args.namespace.resolve()),backup_after=str(args.backup_after.resolve()),
        backup_before=str(args.backup_before.resolve()) if args.backup_before else None,
        previous_report=str(args.previous_report.resolve()),previous_report_sha256=args.previous_sha256,
        previous_prefix=prior,before_receipt_sha256=before_sha,after_receipt_sha256=after_sha,
        closed_prefix=prefix,total_entries=N,chunk_count=len(names)-1,file_count=len(names),total_bytes=total_bytes,
        live_records=totals['live_records'],holes=totals['holes'],closed_representatives=totals['closed_representatives'],
        value_checksum=totals['value_checksum'],new_chunks=len(new_rows),new_indices=prefix-prior,
        new_live_records=totals['live_records']-totals['prior_live_records'],
        new_closed_representatives=totals['closed_representatives']-totals['prior_closed_representatives'],
        new_value_checksum=(totals['value_checksum']-totals['prior_value_checksum'])&MASK,
        aliases_to_closed_representatives=totals['aliases_to_closed_representatives'],
        aliases_to_future_uncomputed_representatives=totals['aliases_to_future_uncomputed_representatives'],
        global_alias_closure_checked=True,complete_domain_closed=prefix==N and totals['aliases_to_future_uncomputed_representatives']==0,
        prior_prefix_files_unchanged=True,before_backup_matches_prior=not args.qualification_replay,
        after_backup_matches_current=True,representative_bitset_bytes=totals['representative_bitset_bytes'],
        representative_bitset_sha256=totals['representative_bitset_sha256'],repair_native_key=key,repair_stabilizer=stab,
        repair_hash=repair_hash,source_header_hash=manifest['source_header_hash'],source_sha256_lineage=PINNED_SHA,
        source_full_catalogue_read=False,global_runtime_extrapolation=False,full_domain_runtime_measured=False,
        worker_sha256=worker_sha,worker_command=command,plan_sha256=plan_sha,stdout_sha256=stdout_sha,
        worker_elapsed_seconds=totals['elapsed_seconds'],worker_peak_rss_bytes=totals['peak_rss_bytes'],
        elapsed_seconds=time.monotonic()-guard.started,parent_peak_rss_bytes=guard.peak,
        engine_summary=summary,files_audited=reports)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    for name in ('namespace','backup-after','previous-report','worker','output'):
        parser.add_argument('--'+name,type=Path,required=True)
    for name in ('backup-before','stdout'):
        parser.add_argument('--'+name,type=Path)
    parser.add_argument('--previous-sha256',required=True)
    parser.add_argument('--expected-prefix',type=int,required=True)
    parser.add_argument('--max-bytes',type=int,required=True)
    parser.add_argument('--maxseconds',type=int,default=180)
    parser.add_argument('--qualification-replay',action='store_true')
    parser.add_argument('--protocol-test',action='store_true',
                        help='execute all production checks, but emit only TEST_ONLY_PASS')
    args = parser.parse_args()
    args.previous_sha256 = args.previous_sha256.upper()
    require(SHA.fullmatch(args.previous_sha256) and 0 < args.expected_prefix <= N and
            (args.expected_prefix % CHUNK == 0 or args.expected_prefix == N) and
            0 < args.max_bytes <= MAX_BYTES and 0 < args.maxseconds <= 180, 'invalid explicit audit bounds')
    require(args.output.suffix == '.json' and args.output.parent.is_dir() and
            args.output.resolve().is_relative_to(LOG_ROOT.resolve()), 'fresh output must be JSON under data/logs')
    for suffix in ('.json','.plan.txt','.worker.jsonl','.worker.stderr.txt'):
        require(not args.output.with_suffix(suffix).exists(), 'audit artifact must be fresh')
    guard = HardBounds(args.maxseconds,1<<30)  # Parent + worker each capped at1GiB; normally far smaller.
    try:
        report = audit(args,guard)
        write_json(args.output,report)
        print(json.dumps({k:v for k,v in report.items() if k not in ('files_audited','engine_summary','worker_command')},indent=2))
        print('[OK] independent two-pass bitset audit; source checkpoints untouched')
    finally:
        guard.close()


if __name__ == '__main__':
    main()

File: experiments/proto/layer_shared_bitset_audit.cpp
// Independent byte/alias auditor. No producer headers or counting algorithms.
// Build: g++ -O3 -mpopcnt -std=c++20 this.cpp -o build/...exe -lbcrypt -lpsapi
#define NOMINMAX
#include <windows.h>
#include <bcrypt.h>
#include <psapi.h>
#include <algorithm>
#include <array>
#include <atomic>
#include <chrono>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>
using U64 = uint64_t;
namespace fs = std::filesystem;
static constexpr U64 SEED=0x5344464a434b3031ULL, N=903398621, CHUNK=25000;
static constexpr U64 SOURCE_HEADER=10599038447932118899ULL, REPAIR=8375264605690466664ULL;
static const std::string SOURCE="ECF0837315B0FDF8AE21C394FDA6676490E43B1828A0825529A344EC17E4E844";
static constexpr U64 MAX_BYTES=12*N+256*((N+CHUNK-1)/CHUNK+1);
static void need(bool ok,const std::string& s){if(!ok)throw std::runtime_error(s);}
static U64 number(const std::string& s){
    need(!s.empty()&&s.size()<=20,"invalid unsigned integer"); U64 n=0;
    for(char c:s){need(c>='0'&&c<='9',"invalid unsigned integer");
        need(n<=(UINT64_MAX-U64(c-'0'))/10,"unsigned integer overflow");n=n*10+U64(c-'0');}
    return n;
}
static U64 le(const uint8_t*p,size_t n){U64 v=0;for(size_t i=0;i<n;++i)v|=U64(p[i])<<(8*i);return v;}
static U64 ck(const uint8_t*p,size_t n,U64 seed=SEED){
    U64 h=seed^(0x9e3779b97f4a7c15ULL+n);size_t i=0;
    for(;i+8<=n;i+=8){h^=le(p+i,8);h*=0xff51afd7ed558ccdULL;h^=h>>33;}
    h^=le(p+i,n-i);h*=0xc4ceb9fe1a85ec53ULL;return h^(h>>33);
}
static bool hexsha(const std::string&s){return s.size()==64&&std::all_of(s.begin(),s.end(),[](char c){return(c>='0'&&c<='9')||(c>='A'&&c<='F');});}
class Sha {
    BCRYPT_ALG_HANDLE algorithm=nullptr;
public:
    Sha(){need(BCryptOpenAlgorithmProvider(&algorithm,BCRYPT_SHA256_ALGORITHM,nullptr,0)>=0,"SHA provider failure");}
    ~Sha(){if(algorithm)BCryptCloseAlgorithmProvider(algorithm,0);}
    std::string operator()(const uint8_t*p,size_t n){
        need(n<=ULONG_MAX,"SHA bounded buffer overflow");BCRYPT_HASH_HANDLE h=nullptr;
        need(BCryptCreateHash(algorithm,&h,nullptr,0,nullptr,0,0)>=0,"SHA create failure");
        std::array<uint8_t,32> out{};
        auto a=BCryptHashData(h,const_cast<PUCHAR>(p),ULONG(n),0);
        auto b=BCryptFinishHash(h,out.data(),ULONG(out.size()),0);BCryptDestroyHash(h);
        need(a>=0&&b>=0,"SHA computation failure");
        const char*d="0123456789ABCDEF";std::string s;for(uint8_t v:out){s+=d[v>>4];s+=d[v&15];}return s;
    }
    std::string operator()(const std::vector<uint8_t>& v){return (*this)(v.data(),v.size());}
};
static std::vector<uint8_t> read(const fs::path&p,U64 expected,U64 maximum){
    need(expected<=maximum,"read exceeds declared byte bound");
    HANDLE h=CreateFileW(p.c_str(),GENERIC_READ,FILE_SHARE_READ,nullptr,OPEN_EXISTING,
        FILE_FLAG_SEQUENTIAL_SCAN|FILE_FLAG_OPEN_REPARSE_POINT,nullptr);
    need(h!=INVALID_HANDLE_VALUE,"cannot open immutable input: "+p.string());
    struct Close{HANDLE h;~Close(){CloseHandle(h);}}close{h};
    BY_HANDLE_FILE_INFORMATION info{};LARGE_INTEGER size{};
    need(GetFileInformationByHandle(h,&info)&&!(info.dwFileAttributes&(FILE_ATTRIBUTE_REPARSE_POINT|FILE_ATTRIBUTE_DIRECTORY)),"input is not an ordinary file");
    need(GetFileSizeEx(h,&size)&&size.QuadPart>=0&&U64(size.QuadPart)==expected,"input byte length differs: "+p.string());
    std::vector<uint8_t> v(static_cast<size_t>(expected));size_t at=0;
    while(at<v.size()){DWORD got=0,want=DWORD(std::min<size_t>(v.size()-at,1<<20));
        need(ReadFile(h,v.data()+at,want,&got,nullptr)&&got>0,"short input read");at+=got;}
    uint8_t extra;DWORD got=0;need(ReadFile(h,&extra,1,&got,nullptr)&&got==0,"input has trailing bytes");return v;
}
struct Header {U64 begin,count,payload,reps,live,sum;bool manifest;};
static Header decode(const std::vector<uint8_t>&v){
    need(v.size()>=256,"truncated header");
    std::string magic(reinterpret_cast<const char*>(v.data()),8);
    need(magic=="SFR4CK01"||magic=="SFR4MT01","wrong shared header magic");
    need(le(v.data()+8,4)==1&&le(v.data()+12,4)==6,"wrong header version/domain");
    need(std::string(reinterpret_cast<const char*>(v.data()+16),64)==SOURCE,"unrecognized source SHA lineage");
    need(le(v.data()+80,8)==0x31304e494d344653ULL&&le(v.data()+88,8)==SOURCE_HEADER&&
         le(v.data()+96,8)==REPAIR&&le(v.data()+104,8)==N&&le(v.data()+112,8)==CHUNK,
         "wrong production semantics/source/repair/domain lineage");
    need(std::all_of(v.begin()+176,v.begin()+256,[](uint8_t x){return x==0;}),"nonzero header reserved bytes");
    std::array<uint8_t,256>b;std::copy_n(v.begin(),256,b.begin());std::fill(b.begin()+144,b.begin()+152,0);
    need(ck(b.data(),b.size())==le(v.data()+144,8),"header checksum mismatch");
    Header h{le(v.data()+120,8),le(v.data()+128,8),le(v.data()+136,8),
        le(v.data()+152,8),le(v.data()+160,8),le(v.data()+168,8),magic=="SFR4MT01"};
    if(h.manifest)need(v.size()==256&&!(h.begin|h.count|h.payload|h.reps|h.live|h.sum),"manifest contains record data");
    return h;
}
struct Row {std::string name,sha,payloadSha;U64 bytes=0,begin=0,count=0,live=0,reps=0,sum=0;bool old=false;};
static std::string chunkName(U64 begin){std::ostringstream s;s<<"chunk-"<<std::hex<<std::setw(16)<<std::setfill('0')<<begin<<".bin";return s.str();}
static std::vector<std::string> tokens(const std::string&line,size_t n){
    std::istringstream s(line);std::vector<std::string> v;std::string x;while(s>>x)v.push_back(x);
    need(v.size()==n,"malformed plan row/token count");return v;
}
int main(int argc,char**argv){
    std::atomic<bool> done=false;std::thread watchdog;
    try{
        std::map<std::string,std::string>a;
        for(int i=1;i<argc;++i){std::string s=argv[i];auto p=s.find('=');
            need(p!=std::string::npos&&p&&p+1<s.size(),"expected key=value option");
            need(a.emplace(s.substr(0,p),s.substr(p+1)).second,"duplicate option");}
        const std::vector<std::string> keys={"plan","plan_sha256","current","after","before","maxseconds","maxrssmib"};
        need(a.size()==keys.size(),"wrong option count");for(const auto&k:keys)need(a.count(k),"missing option: "+k);
        U64 seconds=number(a.at("maxseconds")),rss=number(a.at("maxrssmib"));
        need(seconds>0&&seconds<=180&&rss>=128&&rss<=1024,"invalid hard time/RSS bound");
        auto started=std::chrono::steady_clock::now();
        watchdog=std::thread([&]{while(!done.load()){
            PROCESS_MEMORY_COUNTERS p{};p.cb=sizeof(p);
            if(!GetProcessMemoryInfo(GetCurrentProcess(),&p,sizeof(p))||p.WorkingSetSize>(rss<<20)||
               std::chrono::duration<double>(std::chrono::steady_clock::now()-started).count()>seconds){
                std::cerr<<"BOUND independent native audit time/RSS; no completed report accepted\n";std::cerr.flush();ExitProcess(124);}
            std::this_thread::sleep_for(std::chrono::milliseconds(20));}});
        Sha sha;fs::path plan=fs::u8path(a.at("plan"));
        auto rawPlan=read(plan,fs::file_size(plan),16<<20);
        need(hexsha(a.at("plan_sha256"))&&sha(rawPlan)==a.at("plan_sha256"),"plan SHA mismatch");
        std::string text(rawPlan.begin(),rawPlan.end());need(text.find('\0')==std::string::npos,"NUL in plan");
        std::istringstream input(text);std::string line;need(bool(std::getline(input,line))&&line=="SHARED_F4_BITSET_PLAN_V1","wrong plan schema");
        need(bool(std::getline(input,line)),"missing plan geometry");auto p=tokens(line,5);
        U64 prefix=number(p[0]),prior=number(p[1]),fileCount=number(p[2]),byteBound=number(p[3]);
        bool replay=p[4]=="replay";need(replay||p[4]=="resume","invalid plan mode");
        need(prefix>0&&prefix<=N&&(prefix%CHUNK==0||prefix==N),"invalid prefix boundary");
        need(prior>0&&prior<=prefix&&(replay?prior==prefix:prior<prefix)&&
             (prior%CHUNK==0||prior==N),"invalid predecessor prefix");
        need(fileCount==(prefix+CHUNK-1)/CHUNK+1&&byteBound>0&&byteBound<=MAX_BYTES,"invalid plan bounds");
        std::vector<Row> rows;U64 bytes=0;
        for(U64 i=0;i<fileCount;++i){need(bool(std::getline(input,line)),"truncated plan");auto t=tokens(line,4);
            Row r;r.name=t[0];r.bytes=number(t[1]);r.sha=t[2];r.old=t[3]=="1";
            need((r.old||t[3]=="0")&&hexsha(r.sha),"invalid plan old flag/SHA");
            U64 begin=i*CHUNK;bool manifest=i+1==fileCount;
            U64 size=manifest?256:256+12*std::min(CHUNK,N-begin);
            need(r.name==(manifest?"manifest.bin":chunkName(begin))&&r.bytes==size,"plan filename/range/length mismatch");
            need(r.old==(!replay&&(manifest||begin<prior)),"plan prior-copy flag differs from exact prefix");
            bytes+=r.bytes;rows.push_back(r);
        }
        need(!std::getline(input,line),"trailing plan data");need(bytes<=byteBound,"planned bytes exceed bound");
        fs::path current=fs::u8path(a.at("current")),after=fs::u8path(a.at("after")),before=fs::u8path(a.at("before"));
        std::vector<uint8_t> representatives(size_t((prefix+7)/8),0);
        U64 live=0,reps=0,sum=0,priorLive=0,priorReps=0,priorSum=0;
        for(auto&r:rows){
            auto raw=read(current/r.name,r.bytes,300256);need(sha(raw)==r.sha,"current file SHA mismatch: "+r.name);
            auto copy=read(after/r.name,r.bytes,300256);need(copy==raw&&sha(copy)==r.sha,"after physical copy differs: "+r.name);
            if(r.old){copy=read(before/r.name,r.bytes,300256);need(copy==raw&&sha(copy)==r.sha,"before physical copy differs: "+r.name);}
            Header h=decode(raw);if(r.name=="manifest.bin"){need(h.manifest,"manifest name/magic mismatch");continue;}
            need(!h.manifest&&r.name==chunkName(h.begin)&&h.count==std::min(CHUNK,N-h.begin)&&
                 h.begin<prefix&&h.count<=prefix-h.begin,"chunk range mismatch");
            const uint8_t*aliases=raw.data()+256;const uint8_t*values=aliases+4*h.count;
            need(ck(values,size_t(8*h.count),ck(aliases,size_t(4*h.count)))==h.payload,"payload checksum mismatch");
            r.begin=h.begin;r.count=h.count;r.payloadSha=sha(raw.data()+256,raw.size()-256);
            for(U64 i=0;i<h.count;++i){U64 alias=le(aliases+4*i,4),value=le(values+8*i,8),id=h.begin+i;
                if(alias==UINT32_MAX){need(value==0,"hole stores F4");continue;}
                need(alias<N,"alias outside full native domain");++r.live;
                if(alias==id){need(value>0&&value%24==0,"self representative has invalid F4");
                    ++r.reps;r.sum+=value;representatives[size_t(id>>3)]|=uint8_t(1u<<(id&7));}
                else need(value==0,"nonrepresentative stores F4");
            }
            need(r.live==h.live&&r.reps==h.reps&&r.sum==h.sum,"decoded counters differ from header");
            live+=r.live;reps+=r.reps;sum+=r.sum;
            if(h.begin<prior){priorLive+=r.live;priorReps+=r.reps;priorSum+=r.sum;}
        }
        U64 resolved=0,future=0,holes=0;
        for(const auto&r:rows){auto raw=read(current/r.name,r.bytes,300256);
            need(sha(raw)==r.sha,"current file changed between passes: "+r.name);
            if(r.name=="manifest.bin")continue;
            for(U64 i=0;i<r.count;++i){U64 alias=le(raw.data()+256+4*i,4);
                if(alias==UINT32_MAX){++holes;continue;}
                if(alias<prefix){need(representatives[size_t(alias>>3)]&(1u<<(alias&7)),"alias into committed prefix is not a closed representative");++resolved;}
                else{need(alias<N,"second-pass alias outside full native domain");++future;}
            }
        }
        need(live+holes==prefix&&resolved+future==live,"closure counters inconsistent");
        for(const auto&r:rows){std::cout<<"{\"name\":\""<<r.name<<"\",\"bytes\":"<<r.bytes<<",\"sha256\":\""<<r.sha<<"\"";
            if(r.name!="manifest.bin")std::cout<<",\"begin\":"<<r.begin<<",\"count\":"<<r.count<<",\"live\":"<<r.live<<",\"representatives\":"<<r.reps<<",\"value_checksum\":"<<r.sum<<",\"payload_sha256\":\""<<r.payloadSha<<"\"";
            std::cout<<"}\n";}
        PROCESS_MEMORY_COUNTERS memory{};memory.cb=sizeof(memory);need(GetProcessMemoryInfo(GetCurrentProcess(),&memory,sizeof(memory)),"cannot query final RSS");
        std::cout<<"{\"worker_status\":\"PASS\",\"closed_prefix\":"<<prefix<<",\"total_bytes\":"<<bytes
          <<",\"live_records\":"<<live<<",\"holes\":"<<holes<<",\"closed_representatives\":"<<reps<<",\"value_checksum\":"<<sum
          <<",\"prior_live_records\":"<<priorLive<<",\"prior_closed_representatives\":"<<priorReps<<",\"prior_value_checksum\":"<<priorSum
          <<",\"aliases_to_closed_representatives\":"<<resolved<<",\"aliases_to_future_uncomputed_representatives\":"<<future
          <<",\"representative_bitset_bytes\":"<<representatives.size()<<",\"representative_bitset_sha256\":\""<<sha(representatives)<<"\""
          <<",\"elapsed_seconds\":"<<std::chrono::duration<double>(std::chrono::steady_clock::now()-started).count()
          <<",\"peak_rss_bytes\":"<<memory.PeakWorkingSetSize<<"}\n";
        std::cout.flush();need(bool(std::cout),"worker output failed");done=true;watchdog.join();return 0;
    }catch(const std::exception&e){done=true;if(watchdog.joinable())watchdog.join();std::cerr<<"FAIL "<<e.what()<<"\n";return 1;}
}
